HACKING.md

```diff
@@ -180,26 +180,71 @@ next to the relevant content (e.g. headings, etc.).
 
 Noteworthy changes should also get a new entry in [CHANGES.md](CHANGES.md).
 
+To update the requirements list for the documentation, ensure that
+[pip-tools](https://pypi.org/project/pip-tools/) is installed and run, in the
+root project folder:
 
-## Dependencies
+```sh
+pip-compile --extra=docs -o docs/requirements.txt pyproject.toml
```

Comment:
It's only required when you actually want to bump a dependency, but yes a tool is likely better.

We could have it like this instead, avoiding any lock file at all:

```
[testenv:docs]
description = Build the Sphinx docs
deps = .[docs]
commands =
    sphinx-build -W --keep-going docs/ docs/_build/html
```

The biggest issue is that if the dependencies change, pip doesn't pick up on that, and you generally have to blow away the .tox/env folder, which is not much better than running a command.

We could have the dev dependencies in `tox.ini`, particularly since they're being spilt out, and especially if we don't group many of them (or if there is a way to say "use this same dependency list for multiple envs", other than having them in another file). One advantage of having them in standard places is that tools (like security scanners) easily find them, though, and I'm not sure if they would there.

Or I can just do as you suggest and we keep the generated files (probably not in version control, per a different thread) and I can add a `tox -e update-deps` (name to be bikesheded later).

Reply:
Just having the `.[unit]` style `deps` value in `tox.ini` is no good - it avoids needing any compilation of what's in `pyproject.toml` but the failure to pick up on changes is too inconvenient.

In terms of speed (I did these 3 times and picked the median - not super accurate, but should be enough to spot order-of-magnitude differences), for a full `tox`:

| | Clean | Existing env |
| - | ----- | ------------ |
| current main | 1m47s | 1m15s |
| compile & sync | 1m38s | 1m27s |
| sync only | 1m23s | 1m12s |
| deps-in-tox.ini | 1m19s | 1m12s |

However, most of the time is (thankfully!) in the tests. For a quick env, like `lint`:

| | Clean | Existing env |
| - | ----- | ------------ |
| current main | 13s | 3s |
| compile & sync | 11s | 8s |
| sync only | 6s | 3s |
| deps-in-tox.ini | 5s | 3s |

I did some more investigation, and dependabot will find dependencies in `pyproject.toml` without having a lock file committed, but [won't find them in tox.ini](https://github.com/dependabot/dependabot-core/issues/2076). Other tools seem mixed, but as far as I know we aren't using any at the moment, so I guess we could address that when/if that changes.

I don't want to be significantly slower than current main. So that eliminates having both the `pip-compile` and `pip-sync` in `tox.ini`.

Having the dependencies in `optional-dependencies` sections in `pyproject.toml` is recommended in some places and in others it seems like that's used only for optional features rather than dev dependencies. Being able to do a local install with `pip install .[static,unit]` is nice, but it seems like the vast majority of the time people will be using `tox`, and doing `pip install ops[unit]` doesn't seem like it's of use. Having requirements files generated seems inconvenient if we're not using them for anything else.

So, on balance, I think putting them directly in `tox.ini` is the cleanest for now, and we can reconsider this if the scanning issue ever comes up.